### PR TITLE
Rfs 225 - optional fork activation for regtest and devnet

### PR DESCRIPTION
**Pull request by josedahlquist**
Before proceeding with the first fork, we should leave a way to determine whether the forks are automatically active in regtest/devnet or if they should be activated at certain height. This last capability is quite useful for running integration tests.

I'm adding a new configuration:
```
blockchain.config.hardforkActivationHeights = {
   firstFork = ###
}
```
This configuration is completely optional, and will be considered only for the mentioned networks (regtest and devnet). By default these network will use the last available fork config.

Please let me know if the chosen name is not correct.

The config will change when we define the fork name.

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 12 issues
* ![CRITICAL][CRITICAL] 5 critical
* ![MAJOR][MAJOR] 4 major
* ![INFO][INFO] 3 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [SystemProperties.java#L27](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java#L27): Unused import - org.ethereum.config.blockchain.regtest.RegTestFirstForkConfig. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [DevNetFirstForkConfig.java#L22](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/blockchain/devnet/DevNetFirstForkConfig.java#L22): Unused import - co.rsk.config.BridgeConstants. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [DevNetFirstForkConfig.java#L23](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/blockchain/devnet/DevNetFirstForkConfig.java#L23): Unused import - co.rsk.config.BridgeDevNetConstants. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [DevNetFirstForkConfig.java#L24](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/blockchain/devnet/DevNetFirstForkConfig.java#L24): Unused import - org.ethereum.config.blockchain.testnet.TestNetAfterBridgeSyncConfig. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [AbstractNetConfig.java#L26](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/net/AbstractNetConfig.java#L26): Unused import - java.util.Arrays. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![MAJOR][MAJOR] [DevNetFirstForkConfig.java#L31](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/blockchain/devnet/DevNetFirstForkConfig.java#L31): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [RegTestFirstForkConfig.java#L24](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/blockchain/regtest/RegTestFirstForkConfig.java#L24): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [DevNetConfig.java#L31](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/net/DevNetConfig.java#L31): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [RegTestConfig.java#L30](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/net/RegTestConfig.java#L30): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![INFO][INFO] [SystemProperties.java#L74](https://github.com/rsksmart/rskj/blob/d58a704090893575b333054f1046507a24529bcc/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java#L74): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by tinchou** [comment]
Try to avoid using `config.hasPath`, and instead provide sane defaults in `reference.conf`

**Comment by rskops** [comment]
SonarQube analysis reported 3 issues

1. ![INFO][INFO] [SystemProperties.java#L73](https://github.com/rsksmart/rskj/blob/051c8142f8f800b4815e61986655f78f51937ee9/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java#L73): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)
1. ![INFO][INFO] [HardForkActivationConfig.java#L8](https://github.com/rsksmart/rskj/blob/051c8142f8f800b4815e61986655f78f51937ee9/rskj-core/src/main/java/org/ethereum/config/blockchain/HardForkActivationConfig.java#L8): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)
1. ![INFO][INFO] [RegTestFirstForkConfig.java#L26](https://github.com/rsksmart/rskj/blob/051c8142f8f800b4815e61986655f78f51937ee9/rskj-core/src/main/java/org/ethereum/config/blockchain/regtest/RegTestFirstForkConfig.java#L26): Method org.ethereum.config.blockchain.regtest.RegTestFirstForkConfig.areBridgeTxsFree() is implemented with an exact copy of its superclass's method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACOM_COPIED_OVERRIDDEN_METHOD)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by rskops** [comment]
SonarQube analysis reported 2 issues

1. ![INFO][INFO] [SystemProperties.java#L73](https://github.com/rsksmart/rskj/blob/67d4d3bd6d0a50a063df1221744cb960d80c9923/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java#L73): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)
1. ![INFO][INFO] [HardForkActivationConfig.java#L8](https://github.com/rsksmart/rskj/blob/67d4d3bd6d0a50a063df1221744cb960d80c9923/rskj-core/src/main/java/org/ethereum/config/blockchain/HardForkActivationConfig.java#L8): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: rskj-core/src/main/java/org/ethereum/config/blockchain/HardForkActivationConfig.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+public class HardForkActivationConfig {
+    private final int firstFork;
+
+    // TODO: define a proper name for this config setting
+    private static final String PROPERTY_FIRST_FORK_NAME = "firstFork";
+
+    public HardForkActivationConfig(Config config) {
+        // If I don't have any config for firstFork I will set it to 0
+        this.firstFork = config.hasPath(PROPERTY_FIRST_FORK_NAME) ? config.getInt(PROPERTY_FIRST_FORK_NAME) : 0;
```
Comment (amendelzon) [inline rskj-core/src/main/java/org/ethereum/config/blockchain/HardForkActivationConfig.java]: Maybe define a constant for the default value?

**Pull request merged**